### PR TITLE
[url_launcher] Add non-browser URL launch mode for Android

### PR DESCRIPTION
**Pull request by Mosc**
Adds support for `LaunchMode.externalNonBrowserApplication` to `url_launcher` on Android by acting on the `universalLinksOnly` parameter that was already being passed to the method channel. Simply adds `Intent.FLAG_ACTIVITY_REQUIRE_NON_BROWSER` on Android R and up, but requires a fallback on lower versions. The fallback works by querying the package manager for which apps can handle the intent, and only considering those apps that don't handle just _any_ HTTPS URL. If the resulting list is non-empty, a suitable non-browser app has been found.

Fixes https://github.com/flutter/flutter/issues/66721.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/main/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/url_launcher/url_launcher/lib/src/url_launcher_uri.dart [hunk 1]
```diff
@@ -25,9 +25,9 @@ import 'package:url_launcher_platform_interface/url_launcher_platform_interface.
 ///     On iOS, this should be used in cases where sharing the cookies of the
 ///     user's browser is important, such as SSO flows, since Safari View
 ///     Controller does not share the browser's context.
-///   - [LaunchMode.externalNonBrowserApplication] is supported on iOS 10+.
-///     This setting is used to require universal links to open in a non-browser
-///     application.
+///   - [LaunchMode.externalNonBrowserApplication] is supported on iOS 10+ and
+///     Android. This setting is used to require universal links to open in a
```
Comment (stuartmorgan) [inline packages/url_launcher/url_launcher/lib/src/url_launcher_uri.dart:29]: This change will need to be a separate PR, since you need to update the minimum Android implementation package version to say this.

**Review by stuartmorgan — COMMENTED** [review]
Thanks for the submission! We'll need to hold off briefly on reviewing this, since we should do https://github.com/flutter/plugins/pull/5966 first. That will allow implementing this for the new functionality without changing the legacy API's behavior on Android (which we don't want to do since that legacy API is very iOS-centric). It will also mean that the Android implementation doesn't need to use the iOS term internally, which may be confusing for maintenance.

**Comment by stuartmorgan** [comment]
The PR referenced above has landed, so you can update this to implement the new API now. (It won't currently work from the app-facing package, but that will be resolved shortly with https://github.com/flutter/plugins/pull/5985 so development of this can proceed in parallel.)

**Comment by stuartmorgan** [comment]
The app-facing package PR has also landed, so an updated version of this would work end to end.

**Comment by Mosc** [comment]
Yes, but it's bit low on my priority list, and with limited time available I haven't got around to restructuring the PR yet. I would not mind if you or someone else takes over in the meantime.

**Comment by Hixie** [comment]
@Mosc I'm going to close this PR for now to get it off our review queue, but please don't hesitate to submit a new PR if you have the time to get back to it. In the meantime I'll mention this PR on the issue so that anyone else who wants to pick it up can do so. Thanks again for your contribution!
